engine/util/rng.hpp

```diff
@@ -194,66 +420,82 @@ double basic_rng_t<Engine>::range( double min, double max )
-    {
-      z = gauss_pair_value;
-      gauss_pair_use = false;
-    }
-    else
-    {
-      double x1, x2, w, y1, y2;
-      do
-      {
-        x1 = 2.0 * real() - 1.0;
-        x2 = 2.0 * real() - 1.0;
-        w = x1 * x1 + x2 * x2;
-      }
-      while ( w >= 1.0 || w == 0.0 );
-
-      w = std::sqrt( ( -2.0 * std::log( w ) ) / w );
-      y1 = x1 * w;
-      y2 = x2 * w;
-
-      z = y1;
-      gauss_pair_value = y2;
-      gauss_pair_use = true;
-    }
+    gauss_pair_use = false;
+    return mean + gauss_pair_value * stddev;
   }
-  else
-    z = 0.0;
 
-  double result = mean + z * stddev;
+  double x1, x2, w, y1, y2;
+  do
+  {
+    x1 = 2.0 * real() - 1.0;
+    x2 = 2.0 * real() - 1.0;
+    w = x1 * x1 + x2 * x2;
+  }
+  while ( w >= 1.0 || w == 0.0 );
+
+  w = std::sqrt( ( -2.0 * std::log( w ) ) / w );
+  y1 = x1 * w;
+  y2 = x2 * w;
 
-  // True gaussian distribution can of course yield any number at some probability.  So truncate on the low end.
-  if ( truncate_low_end && result < 0 )
-    result = 0;
+  gauss_pair_value = y2;
+  gauss_pair_use = true;
 
-  return result;
+  return mean + y1 * stddev;
+}
+
+template <typename Engine>
+double basic_rng_t<Engine>::gauss_ab( double mean, double stddev, double min, double max )
+{
+  assert( stddev >= 0.0 && "Stddev must be non-negative." );
+  assert( min < max && "Minimum must be less than maximum." );
+  assert( mean >= min && mean <= max && "Mean must be contained within the interval [min, max]" );
+
+  if ( stddev == 0.0 )
+    return mean;
+
+  double min_cdf      = stdnormal_cdf( ( min - mean ) / stddev );
+  double max_cdf      = stdnormal_cdf( ( max - mean ) / stddev );
+  double uniform      = real();
+  double rescaled_cdf = min_cdf + uniform * ( max_cdf - min_cdf );
+  return mean + stddev * stdnormal_inv( rescaled_cdf );
+}
+
+template <typename Engine>
+double basic_rng_t<Engine>::gauss_a( double mean, double stddev, double min )
+{
+  return gauss_ab( mean, stddev, min, std::numeric_limits<double>::infinity() );
+}
+
+template <typename Engine>
+double basic_rng_t<Engine>::gauss_b( double mean, double stddev, double max )
+{
+  return gauss_ab( mean, stddev, -std::numeric_limits<double>::infinity(), max );
 }
 
 /// Exponential Distribution
 template <typename Engine>
 double basic_rng_t<Engine>::exponential( double nu )
 {
   double x;
-  do { x = real(); } while ( x >= 1.0 ); // avoid ln(0)
+  do { x = real(); } while ( x >= 1.0 ); // avoid ln(k) where k <= 0. this should be guaranteed by `real()`, but just in case
+  // TODO: check if `real()` generates any non-positive values
   return - std::log( 1 - x ) * nu;
 }
 
 /// Exponentially Modified Gaussian Distribution
 template <typename Engine>
 double basic_rng_t<Engine>::exgauss( double gauss_mean, double gauss_stddev, double exp_nu )
 {
-  double v = gauss( gauss_mean, gauss_stddev ) + exponential( exp_nu );
-  return v > 0.0 ? v : 0.0;
+  return gauss( gauss_mean, gauss_stddev ) + exponential( exp_nu );
```

Comment:
player_t::total_reaction_time should be changed to ensure no negative values can happen, although with the default parameters the probability of that is extremely low.

Reply:
I actually need to verify some behaviour on this front, but I'll address it tomorrow.

Reply:
After further review, I see two potential solutions:
- Provide an overload for unsigned parameter inputs in which `gauss_a( gauss_mean, gauss_stddev, 0.0)` is called instead of `gauss(...)`.
- Opt for `gauss_a`'s usage in `exgauss` instead of `gauss` altogether (no conditioning on signedness of type).

For consistency's sake, the former seems most reasonable.